### DISCUSSION TITLE
Compiling SCSS source code using serverless functions

### DISCUSSION THREAD

**Discussion by jerryjappinen** [vercel/community#1222] [Help] [answered]
### Summary

I have a CSS library. The source code is written in SCSS and includes dynamic variables (like breakpoint values). I want to compile the final output CSS with the variables I get from user input: I have a web site where a user can enter their values, and I then pass them on to a serverless function. This function compiles the source code using `node-sass`.

It's working locally, but on `vercel.app` the serverless function no longer finds the source files. I'm not sure if this is because the full source code is not available to the serverless function when deployed, or because the way I get the path does not work on production.

### Additional information

The source files are included in the same repo. My code structure looks like this:

```
api/
  compile.js
source
  mixins/
    some-mixin-used-by-lib.scss
  lib.scss
```

Note that the source code has multiple files, and the `.scss` files import other `.scss` files.

Currently, I'm getting the file path [like this](https://vercel.com/guides/how-can-i-use-files-in-serverless-functions):

```js
const sourceCodePath = path.join(process.cwd() + './' + 'source/lib.scss')
```

As I said, this works as desired on `vercel dev` but not when deployed to Vercel.

What would be the recommended way to make this work?

**Comment by amyegan** [comment]
Hi @jerryjappinen! Do you have a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) that I can use to get a better look at what is going wrong?

**Comment by jerryjappinen** [comment, marked as answer]
@amyegan sure, you can look at the repo here: https://github.com/jerryjappinen/layers-css

I was hesitant to share this up front because the repo is old and messy (an older project I moved over from PHP). The relevant parts are the endpoints under `api/`, which compile the source code from `source/`.

The UI part is shown at [layers-css.vercel.app](https://layers-css.vercel.app/) if it helps, but you can just look at `api/responsive.scss` to see what how the query parameters are handled to get the idea.

**Reply by amyegan** [reply]
Thanks for sharing this! I'm getting the error message `Error: File to import not found or unreadable`. At first, I thought this was related to the relative `@imports` in your `responsive.scss` file. But, after quite a bit more digging, I think it's actually a being caused by `sourcePath` when building the `mixinsPath`.

I was able to fix this a couple of different ways, though I'm sure there are many other possible solutions.

_Option 1:_ Change the `mixinsPath` line to 
```
const mixinsPath = path.join(process.cwd(), "source", "mixins");
```

_Option 2:_ Read in each file to ensure they are available when needed later in the function
```
const mixinsPath = path.join(process.cwd(), sourcePath, 'mixins')
const responsiveMain = fs.readFileSync(path.join(mixinsPath, "responsive.scss"));
const responsiveFirst = fs.readFileSync(path.join(mixinsPath, "responsive-first.scss"));
const responsiveConsecutive = fs.readFileSync(path.join(mixinsPath, "responsive-consecutive.scss"));
```

**Reply by jerryjappinen** [reply]
Thank you for the debugging and answer, @amyegan. I was able to get both endpoints working 🙏

Just in case someone stumbles upon this, and for my own understanding: is the root cause that the files are only there, if Vercel thinks that the script will access them? Or are the files always there after deployment, but the path resolution has to be done in a particular way so that it gets the right path in production?
